packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -148,64 +155,109 @@ export function toVisualTree(tree: VisualSharedTreeNode): VisualChildNode {
+ * Processes and visualizes the fields of a verbose tree node.
+ *
+ * @param treeFields - The fields of the tree node to visualize. Can be either an array of VerboseTree (for array nodes) or a Record of field names to VerboseTree (for object/map nodes).
+ * @param treeDefinitions - Map containing all schema definitions for the entire tree structure. Each definition describes the shape and constraints of a particular node type.
+ * @param allowedTypes - Record mapping field names to their allowed type sets. Specifies which types are valid for each field in the tree.
+ * @param requirements - Optional record mapping field names to boolean values indicating whether each field is required (true) or optional (false). Only meaningful for object node fields.
+ *
+ * @returns A record mapping field names/indices to their visual tree representations.
  */
 async function visualizeVerboseNodeFields(
-	tree: VerboseTreeNode,
-	treeSchema: SimpleTreeSchema,
+	treeFields: readonly VerboseTree[] | Record<string, VerboseTree>,
+	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
+	fieldSchemaProperties: Record<string, FieldSchemaProperties>,
 	visualizeChildData: VisualizeChildData,
 ): Promise<Record<string, VisualSharedTreeNode>> {
-	const treeFields = tree.fields;
-
 	const fields: Record<string | number, VisualSharedTreeNode> = {};
 
 	for (const [fieldKey, childField] of Object.entries(treeFields)) {
-		fields[fieldKey] = await visualizeSharedTreeNodeBySchema(
+		fields[fieldKey] = await visualizeSharedTreeBySchema(
 			childField,
-			treeSchema,
+			treeDefinitions,
+			{
+				allowedTypes: fieldSchemaProperties[fieldKey]?.allowedTypes,
+				isRequired:
+					fieldSchemaProperties[fieldKey]?.isRequired === undefined
+						? undefined
+						: fieldSchemaProperties[fieldKey]?.isRequired,
+			},
 			visualizeChildData,
 		);
 	}
 
 	return fields;
 }
 
+/**
+ * Extracts and stores allowed types & kind from {@link SimpleFieldSchema}.
+ */
+function getFieldTooltipProperties(
+	schema: SimpleObjectNodeSchema,
+): Record<string, FieldSchemaProperties> {
+	const result: Record<string, FieldSchemaProperties> = {};
+
+	for (const [fieldKey, treeFieldSimpleSchema] of Object.entries(schema.fields)) {
+		result[fieldKey] = {
+			allowedTypes: treeFieldSimpleSchema.allowedTypes,
+			isRequired: treeFieldSimpleSchema.kind === FieldKind.Required ? true : false,
+		};
+	}
+
+	return result;
+}
+
 /**
  * Returns the schema & fields of the node with type {@link ObjectNodeStoredSchema}.
  */
 async function visualizeObjectNode(
 	tree: VerboseTreeNode,
-	nodeSchema: SimpleObjectNodeSchema,
-	treeSchema: SimpleTreeSchema,
+	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
+	{ allowedTypes, isRequired }: FieldSchemaProperties,
 	visualizeChildData: VisualizeChildData,
 ): Promise<VisualSharedTreeNode> {
+	const objectNodeSchemaProperties = getFieldTooltipProperties(
+		treeDefinitions.get(tree.type) as SimpleObjectNodeSchema,
+	);
```

Comment:
Is there anything we can do here to guarantee that the `tree` that was passed in will actually result in this returning a `SimpleObjectNodeSchema`? Or there's no way `visualizeObjectNode` can receive something that is an array or map node?

Reply:
In the other functions, it looks like we pass the node schema in as a parameter. Should we do that here too? Then we wouldn't need to do this `treeDefinitions` access / cast.

Reply:
```typescript
async function visualizeObjectNode(
	tree: VerboseTreeNode,
	schema: SimpleObjectNodeSchema,
	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
	{ allowedTypes, isRequired }: FieldSchemaProperties,
	visualizeChildData: VisualizeChildData,
): Promise<VisualSharedTreeNode> {
	const objectNodeSchemaProperties = getFieldTooltipProperties(schema);
```

I changed as above which also alignts witih the `visualizeMapNode()`.